### PR TITLE
ensure objects beyond bump allocated region are inserted into the object pool freelist

### PR DESCRIPTION
**Pull request by d-netto**
x-ref #50137

### CODE REVIEW

**Comment by Seelengrab** [comment]
Please link the PR & comments in question directly, to make it easier for others to follow the train of thought.

File: src/gc.c [hunk 1]
```diff
@@ -1402,7 +1397,9 @@ static jl_taggedvalue_t **gc_sweep_page(jl_gc_pool_t *p, jl_gc_pagemeta_t **allo
         jl_taggedvalue_t **pfl_begin = NULL;
         while ((char*)v <= lim) {
             int bits = v->bits.gc;
-            if (!gc_marked(bits)) {
+            // if an object is past `lim_newpages` then we can guarantee it's garbage
+            if (!gc_marked(bits) || (char*)v >= lim_newpages) {
+                v->bits.gc = GC_CLEAN;
```
Comment (vtjnash) [inline src/gc.c:1402]:
```suggestion
                v->next = NULL; // v is now the tail of the freelist and will be assigned via `*pfl` later
```

**Review by vtjnash — COMMENTED** [review]
LGTM. Any idea if this fixes the recent memory leak claimed?

File: src/gc.c (hunk 1 above)
Comment (d-netto) [reply src/gc.c:1402]: Why is this necessary? Don't we already null out the terminal pointers in https://github.com/JuliaLang/julia/blob/734cafabbfb1040aac7a92682ae78e8a6db699cf/src/gc.c#L1570?

**Comment by inkydragon** [comment]
> Any idea if this fixes the recent memory leak claimed?

Test with `dcn/50137-follow-up/53fe3e3dee` on linux

```
julia> [f(forecast_samples_adj) for i in 1:50] |> diff
49-element Vector{Float64}:
 0.20465087890625
 0.2041015625
 0.2041015625
 0.2041015625
 0.2041015625
 0.2041015625
 0.2041015625
...
```

Seems unchanged, problems still exist.

File: src/gc.c (hunk 1 above)
Comment (vtjnash) [reply src/gc.c:1402]: It is better than the RMW for setting GC_CLEAN, though indeed, this line of code is not needed.
```suggestion
```

**Comment by d-netto** [comment]
I believe the possible memory leak @vtjnash is referring to is not related to https://github.com/JuliaLang/julia/issues/50345, but discussed in this slack thread https://julialang.slack.com/archives/C0309QBEZNH/p1687914365229929.

**Comment by d-netto** [comment]
Any chance we can still backport this to alpha2 @KristofferC (forgot to put the label).

**Comment by d-netto** [comment]
NVM, this PR already made it to alpha1...

**Pull request merged**